pallets/kilt-launch/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+use pallet_vesting::{Vesting, VestingInfo};
+use sp_runtime::traits::{Convert, Saturating};
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+// TODO: Add benchmarking
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+
+pub const KILT_LAUNCH_ID: LockIdentifier = *b"kiltlock";
+pub const VESTING_ID: LockIdentifier = *b"vesting ";
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::pallet_prelude::*;
+
+	#[derive(Debug, Encode, Decode, PartialEq, Eq, Clone)]
+	pub struct LockedBalance<T: Config> {
+		pub block: T::BlockNumber,
+		pub amount: <T as pallet_balances::Config>::Balance,
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_balances::Config + pallet_vesting::Config {
+		/// Because this pallet emits events, it depends on the runtime's
+		/// definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Maximum number of claims which can be migrated in a single call.
+		/// Used for weight estimation.
+
+		/// NOTE:
+		/// + Benchmarks will need to be re-run and weights adjusted if this
+		/// changes. + This pallet assumes that dependents keep to the limit
+		/// without enforcing it.
+		type MaxClaims: Get<usize>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub balance_locks: Vec<(T::AccountId, T::BlockNumber, <T as pallet_balances::Config>::Balance)>,
+		pub transfer_account: T::AccountId,
+		pub vesting: Vec<(T::AccountId, T::BlockNumber, BalanceOf<T>)>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				balance_locks: Default::default(),
+				transfer_account: Default::default(),
+				vesting: Default::default(),
+			}
+		}
+	}
+
+	// Balance type based on pallet_vesting
+	pub type BalanceOf<T> =
+		<<T as pallet_vesting::Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			// Generate initial custom locking configuration
+			// * who - Account which setting the custom lock for
+			// * length - Number of blocks from  until removal of the lock
+			// * amount - Number of tokens which are locked
+			for (ref who, length, locked) in self.balance_locks.iter() {
+				if !length.is_zero() {
+					let balance = <pallet_balances::Pallet<T>>::free_balance(who);
+					assert!(!balance.is_zero(), "Currencies must be init'd before locking");
+					assert!(
+						balance >= *locked,
+						"Locked balance must not exceed total balance for address {:?}",
+						who
+					);
+
+					// Add unlock block to storage
+					<BalanceLocks<T>>::insert(
+						who,
+						LockedBalance::<T> {
+							block: *length,
+							amount: *locked,
+						},
+					);
```

Comment:
Good point. Theoretically, we will go through the list of accounts million times by hand to be sure everything is correct but it will still be good to add a check here for sure.